src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -94,20 +123,30 @@ protected void handleMessageInternal(final Message<?> message) throws Exception
 				? this.messageKeyExpression.getValue(this.evaluationContext, message)
 				: message.getHeaders().get(KafkaHeaders.MESSAGE_KEY);
 
+		ListenableFuture<?> future;
+
 		if (partitionId == null) {
 			if (messageKey == null) {
-				this.kafkaTemplate.send(topic, (V) message.getPayload());
+				future = this.kafkaTemplate.send(topic, (V) message.getPayload());
 			}
 			else {
-				this.kafkaTemplate.send(topic, (K) messageKey, (V) message.getPayload());
+				future = this.kafkaTemplate.send(topic, (K) messageKey, (V) message.getPayload());
 			}
 		}
 		else {
 			if (messageKey == null) {
-				this.kafkaTemplate.send(topic, partitionId, (V) message.getPayload());
+				future = this.kafkaTemplate.send(topic, partitionId, (V) message.getPayload());
+			}
+			else {
+				future = this.kafkaTemplate.send(topic, partitionId, (K) messageKey, (V) message.getPayload());
+			}
+		}
+		if (this.sync) {
+			if (this.sendTimeout < 0) {
+				future.get();
 			}
 			else {
-				this.kafkaTemplate.send(topic, partitionId, (K) messageKey, (V) message.getPayload());
+				future.get(this.sendTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Perhaps we should wrap the `TimeoutException` in a `MessageTimeoutException` with `failedMessage` set to `message`. Otherwise we just get a generic `MessageHandlingException`.